src/trs/hmtrs.rs

```diff
@@ -1,24 +1,82 @@
-use super::utils::{log_n_of, logsumexp};
 ///! Hindley-Milner Typing for First-Order Term Rewriting Systems (no abstraction)
 ///!
 ///! Much thanks to:
 ///! - https://github.com/rob-smallshire/hindley-milner-python
 ///! - https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system
 ///! - (TAPL; Pierce, 2002, ch. 22)
+use super::trace::Trace;
+use super::utils::{log_n_of, logsumexp};
```

Comment:
Should `utils` be incorporated into some more generic utility module for for the entire crate?

Reply:
Yeah we can definitely pull out `logsumexp`. I'll see what I can do about the others